### ISSUE TITLE
Numpy dependencies in pkgs/main/linux-64/repodata.json don't match index.json or recipe

### ISSUE THREAD

**Issue by taldcroft** [ContinuumIO/anaconda-issues#11920]
### Actual Behavior

The repodata (https://repo.anaconda.com/pkgs/main/linux-64/repodata.json) for numpy includes these dependencies:
```
    "numpy-1.18.5-py37ha1c710e_0.tar.bz2": {
      "build": "py37ha1c710e_0",
      "build_number": 0,
      "depends": [
        "blas 1.0 mkl",
        "libgcc-ng >=7.3.0",
        "libgfortran-ng >=7,<8.0a0",
        "mkl >=2019.4,<2021.0a0",   <=== Up through 2021.0a0 allowed
        "mkl-service >=2,<3.0a0",
        "mkl_fft",
        "mkl_random",
        "numpy-base 1.18.5.*",
        "python >=3.7,<3.8.0a0"
      ],
 ```

However, the corresponding recipe (and `index.json`) in `miniconda3/pkgs/numpy-1.18.5-py37ha1c710e_0/info` give these dependencies:
```
 "depends": [
    "blas 1.0 mkl",
    "libgcc-ng >=7.3.0",
    "libgfortran-ng >=7,<8.0a0",
    "mkl >=2019.4,<2020.0a0",  <==== only through 2020.0a0
    "mkl-service >=2,<3.0a0",
    "mkl_fft",
    "mkl_random",
    "numpy-base 1.18.5.*",
    "python >=3.7,<3.8.0a0"
  ],
```

This mismatch causes inconsistencies if I build a local environment and use the downloaded packages to make a local repository. This is needed in order to create a conda environment on a secured computer without network access.

In particular, with the dependency in `repodata.json`, `numpy` installs with `mkl 2020.1`, but I put all those packages into a local repo, run `conda index` and then try to install from that local repo then it fails because `mkl 2020.1` is not allowed.

### Expected Behavior

I expect the dependencies to always match.

### Steps to Reproduce

```
conda create -n test1 numpy python=3.7
conda activate test1

# Copy packages from  "conda list" (in ~/miniconda3/pkgs) appropriately into repo/noarch or repo/linux-64

conda index repo
conda create -n test2 --override-channels -c file://$PWD/repo numpy
# FAIL
```

##### Anaconda or Miniconda version:

conda 4.8.3

##### Operating System:

Linux CentOS-7

##### `conda info`

<details>

```
PASTE OUTPUT HERE:
tmp1) ➜  info conda info

     active environment : tmp1
    active env location : /home/aldcroft/miniconda3/envs/tmp1
            shell level : 7
       user config file : /home/aldcroft/.condarc
 populated config files : 
          conda version : 4.8.3
    conda-build version : 3.18.11
         python version : 3.7.7.final.0
       virtual packages : __glibc=2.17
       base environment : /home/aldcroft/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/aldcroft/miniconda3/pkgs
                          /home/aldcroft/.conda/pkgs
       envs directories : /home/aldcroft/miniconda3/envs
                          /home/aldcroft/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.3 requests/2.23.0 CPython/3.7.7 Linux/3.10.0-862.11.6.el7.x86_64 centos/7.5.1804 glibc/2.17
                UID:GID : 1001:1001
             netrc file : /home/aldcroft/.netrc
           offline mode : False

```

</details>

##### `conda list --show-channel-urls`

<details>

```
PASTE OUTPUT HERE:
(tmp4) ➜  info conda list --show-channel-urls
# packages in environment at /home/aldcroft/miniconda3/envs/tmp4:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
bcrypt                    3.1.7            py37h7b6447c_1    defaults
blas                      1.0                         mkl    defaults
ca-certificates           2020.6.24                     0    defaults
certifi                   2020.6.20                py37_0    defaults
cffi                      1.14.0           py37he30daa8_1    defaults
cryptography              2.9.2            py37h1ba5d50_0    defaults
idna                      2.10                       py_0    defaults
intel-openmp              2019.4                      243    defaults
ld_impl_linux-64          2.33.1               h53a641e_7    defaults
libedit                   3.1.20191231         h7b6447c_0    defaults
libffi                    3.3                  he6710b0_2    defaults
libgcc-ng                 9.1.0                hdf63c60_0    defaults
libgfortran-ng            7.3.0                hdf63c60_0    defaults
libsodium                 1.0.18               h7b6447c_0    defaults
libstdcxx-ng              9.1.0                hdf63c60_0    defaults
mkl                       2019.4                      243    defaults
mkl-service               2.3.0            py37he904b0f_0    defaults
mkl_fft                   1.1.0            py37h23d657b_0    defaults
mkl_random                1.1.0            py37hd6b4f25_0    defaults
ncurses                   6.2                  he6710b0_1    defaults
numpy                     1.18.5           py37ha1c710e_0    defaults
numpy-base                1.18.5           py37hde5b4d6_0    defaults
openssl                   1.1.1g               h7b6447c_0    defaults
paramiko                  2.7.1                      py_0    defaults
pip                       20.1.1                   py37_1    defaults
pycparser                 2.20                       py_0    defaults
pynacl                    1.4.0            py37h7b6447c_0    defaults
python                    3.7.7                hcff3b4d_5    defaults
readline                  8.0                  h7b6447c_0    defaults
setuptools                47.3.1                   py37_0    defaults
six                       1.15.0                     py_0    defaults
sqlite                    3.32.3               h62c20be_0    defaults
tk                        8.6.10               hbc83047_0    defaults
wheel                     0.34.2                   py37_0    defaults
xz                        5.2.5                h7b6447c_0    defaults
zlib                      1.2.11               h7b6447c_3    defaults

```

</details>

**Comment by jjhelmus** [comment]
The dependency information contained in a conda package may not match the information in repodata since repodata is updated as new information about compatibility and incompatibilities between packages are discovered.  Although we try to keep modifications to repodata to a minimum, some patching is needed to ensure that incompatible packages are not installed by conda.

The patches applied to repodata are published in the `patch_instructions.json` file in the package repository for each subdir, for example [for linux-64](https://repo.anaconda.com/pkgs/main/linux-64/patch_instructions.json).  These instructions are generated by `conda index` from the `main.py` script in [AnacondaRecipes/repodata-hotfixes](https://github.com/AnacondaRecipes/repodata-hotfixes/).

If these packages are downloaded for local use it will be necessary to apply these patch instructions to the local repository.  This can be done using the `-p` or `--patch-generator` option of `conda index`.  

As an example, if packages are downloaded to **/home/jhelmus/local_pkgs/** to use the `main.py` script to apply patches:

```
git clone git@github.com:AnacondaRecipes/repodata-hotfixes.git
conda index -p repodata-hotfixes/main.py /home/jhelmus/local_pkgs/
```

To use the precomputed `patch_instructions.json` files:

```
mkdir -p temp/linux-64
curl https://repo.anaconda.com/pkgs/main/linux-64/patch_instructions.json -o temp/linux-64/patch_instructions.json
# ... repeat for any other subdirs used
mkdir -p temp/noarch
curl https://repo.anaconda.com/pkgs/main/noarch/patch_instructions.json -o temp/noarch/patch_instructions.json
tar cf bundle.tar.bz2 -C ./temp .
conda index -p bundle.tar.bz2 /home/jhelmus/local_pkgs/
```

**Comment by taldcroft** [comment]
@jjhelmus - thanks for the explanation, that's very helpful! I had guessed that something like this might be happening. I'm closing this issue as it is not a bug, but I might suggest putting the excellent example you just provided into the conda docs somewhere that might be discoverable with google.

To work around this I had to spend a fair bit of time basically replicating the patch functionality (unpacking .conda packages, fixing `index.json` and repacking). On the plus side I learned a lot about the structure of .conda packages and discovered `libarchive-c` for platform-independent archive manipulation.
